bitcoin/src/consensus/serde.rs

```diff
@@ -0,0 +1,542 @@
+            (Err(ConsensusError::Io(io_error)), de_error) => panic!("Unexpected IO error {:?} returned from {}::consensus_decode(), deserialization error: {:?}", io_error, core::any::type_name::<T>(), de_error),
+            (Err(consensus_error), Some(de_error)) => panic!("{} should've returned `Other` IO error because of deserialization error {:?} but it returned consensus error {:?} instead", core::any::type_name::<T>(), de_error, consensus_error),
+        }
+    }
+}
+
+impl<E: fmt::Debug, I: Iterator<Item=Result<u8, E>>> io::Read for IterReader<E, I> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let mut count = 0;
+        for (dst, src) in buf.iter_mut().zip(&mut self.iterator) {
+            match src {
+                Ok(byte) => *dst = byte,
+                Err(error) => {
+                    self.error = Some(error);
+                    return Err(io::ErrorKind::Other.into());
+                }
+            }
+            // bounded by the length of buf
+            count += 1;
+        }
+        Ok(count)
+    }
+}
+
+/// Helper for `#[serde(with = "")]`.
+///
+/// To (de)serialize a field using consensus encoding you can write e.g.:
+///
+/// ```
+/// use bitcoin::Transaction;
+/// use bitcoin::consensus;
+/// use serde_derive::{Serialize, Deserialize};
+///
+/// #[derive(Serialize, Deserialize)]
+/// # #[serde(crate = "actual_serde")]
+/// pub struct MyStruct {
+///     #[serde(with = "consensus::serde::With::<consensus::serde::Hex>")]
+///     tx: Transaction,
+/// }
+/// ```
+pub struct With<E>(PhantomData<E>);
+
+impl<E> With<E> {
+    /// Serializes the value as consensus-encoded
+    pub fn serialize<T: Encodable, S: Serializer>(value: &T, serializer: S) -> Result<S::Ok, S::Error> where E: ByteEncoder {
+        if serializer.is_human_readable() {
+            serializer.collect_str(&DisplayWrapper::<'_, _, E>(value, Default::default()))
+        } else {
+            use crate::StdError;
+
+            let serializer = serializer.serialize_seq(None)?;
+            let mut writer = BinWriter {
+                serializer,
+                error: None,
+            };
+
+            let result = value.consensus_encode(&mut writer);
+            match (result, writer.error) {
+                (Ok(_), None) => writer.serializer.end(),
+                (Ok(_), Some(error)) => panic!("{} silently ate an IO error: {:?}", core::any::type_name::<T>(), error),
+                (Err(io_error), Some(ser_error)) if io_error.kind() == io::ErrorKind::Other && io_error.source().is_none() => Err(ser_error),
+                (Err(io_error), ser_error) => panic!("{} returned an unexpected IO error: {:?} serialization error: {:?}", core::any::type_name::<T>(), io_error, ser_error),
+            }
+        }
+    }
+
+    /// Deserializes the value as consensus-encoded
+    pub fn deserialize<'d, T: Decodable, D: Deserializer<'d>>(deserializer: D) -> Result<T, D::Error> where for<'a> E: ByteDecoder<'a> {
```

Comment:
That's absolutely not equivalent to `'static`. If you require `'static` then the caller needs `'static` reference, which is not possible here. It also doesn't need to outlive `'d` because if deserializer calls `visit_str` (not to be confused with the borrowed version) we want to be able to deserialize it - this works fine for hex.

Reply:
Unless I am confused, `for<'a> 'a` means "for any `'a`" which is quite hard to achieve without your actual lifetime being `'static`.

If the intention was to qualify with "for some `'a`" then I think you need to add an extra lifetime parameter to the function rather than using `for`.

Reply:
But I may well be confused.

Reply:
It does mean for any `'a` but it actually implies `'static` isn't enough. In practice it means that the `impl` line is `impl<'a> ByteDecoder<'a> for SomeEncoder` and definitely not `impl ByteDecoder<'static> for SomeEncoder`. The struct may or may not have a lifetime.

The reason it is here is we expect the decoder to hold the reference to the string (line 325) and we can't express it without GAT, so we need lifetime parameter on the trait. We also need it to be callable with any lifetime because we receive non-borrowed `str` from serde so we need to locally create a temporary decoder.

We could make the API if we raised MSRV to 1.65 but I guess this will not be ACKed. ;)

Reply:
Ah, now I understand!

I think, if you could add a comment (probably a non-doccomment) saying basically exactly what you said in your first paragraph, that would be helpful. What I was missing is that the arbitrary lifetime was on the trait `impl` rather than the struct, and that the effect is to force `impl<'a> ByteDecoder<'a>` -- meaning that the impl is independent of the lifetime, not that some reference somewhere needs to live forever.

Reply:
I need some sleep to be able to write a good comment. :) I think it shouldn't be critical since things will just work for consumers and implementors will find very quickly that their wrong impl doesn't work.

Reply:
Yeah true, good point :) though for the sake of this review, I was worried that one of us would discover, down the line, that it was impossible not to make a "wrong impl".

Reply:
The possibility of making a good impl is already proven by the code using it. :)